File: app/src/main/java/xyz/mustafaali/devqstiles/service/ToggleUsbDebuggingService.java
package xyz.mustafaali.devqstiles.service;

import android.provider.Settings;
import android.util.Log;
import android.widget.Toast;

import xyz.mustafaali.devqstiles.R;

/**
 * Tile Service to toggle USB Debugging.
 */
public class ToggleUsbDebuggingService extends BaseTileService {

    private final String TAG = this.getClass().getSimpleName();

    @Override
    public void onStartListening() {
        super.onStartListening();
        updateTile();
    }

    @Override
    public void onClick() {
        String newValue = isFeatureEnabled() ? "0" : "1";

        try {
            Settings.Global.putString(contentResolver, Settings.Global.ADB_ENABLED, newValue);
        } catch (SecurityException se) {
            String message = getString(R.string.permission_required_toast);
            Toast.makeText(getApplicationContext(), message, Toast.LENGTH_LONG).show();
            Log.e(TAG, message);
        }
        updateTile();
    }

    @Override
    protected boolean isFeatureEnabled() {
        return Settings.Global.getString(contentResolver, Settings.Global.ADB_ENABLED).equals("1");
    }

}

File: app/src/main/kotlin/xyz/mustafaali/devqstiles/service/ToggleUsbDebuggingService.kt
package xyz.mustafaali.devqstiles.service

import android.provider.Settings
import timber.log.Timber

/**
 * Tile Service to toggle USB Debugging.
 */
class ToggleUsbDebuggingService : BaseTileService() {

    override fun onStartListening() {
        super.onStartListening()
        updateTile()
    }

    override fun onClick() {
        val newValue = if (isFeatureEnabled()) "0" else "1"

        try {
            Settings.Global.putString(contentResolver, Settings.Global.ADB_ENABLED, newValue)
        } catch (se: SecurityException) {
            Timber.e(se, se.message)
            showPermissionError()
        }
        updateTile()
    }

    override fun isFeatureEnabled(): Boolean {
        return Settings.Global.getString(contentResolver, Settings.Global.ADB_ENABLED) == "1"
    }
}
